### ISSUE TITLE
Allowed memory size of 536870912 bytes exhausted

### ISSUE THREAD

**Issue by haheute** [friendica/friendica#12777]
- [x] I have searched open and closed issues for duplicates

### Bug Description

I see many of these Errors in the admin area and don't know what to do with that:

`2023-02-06T22:13:36Z | ALERT | worker | Fatal Error (E_ERROR): Allowed memory size of 536870912 bytes exhausted (tried to allocate 113262592 bytes)`

I'm not doing something special. Just reading the latest posts.. 

For a while I had set the memory limit to 1024M, that did not help, so I changed it back to 512M. 

### Screenshots

![fr-gd-webp](https://user-images.githubusercontent.com/5753419/217100703-73549aa7-b716-44c9-bbac-4a49156e695f.jpg)

### Single Log Entry

![fr-gd-detail](https://user-images.githubusercontent.com/5753419/217101760-a3a56c6f-04be-4fd4-a0b8-c715d96ec4d1.jpg)

### Platform Info

Friendica Version: Friendica 'Giant Rhubarb' 2023.03-dev - 1513

PHP version: 8.1.15

SQL version: mysql  Ver 15.1 Distrib 10.6.11-MariaDB, for Linux (x86_64) using readline 5.1

**Comment by MrPetovan** [comment]
Looks like your node is trying to process a large webp video which hits the memory limit. Too bad we don't know whether the video could be processed by raising the memory limit to 1G or 2G.

**Comment by haheute** [comment]
I am trying with 2048M now.

**Comment by haheute** [comment]
OK, the Error did not occur since yesterday 23 pm (when I changed to 2048M). The other one ist still there (`Fatal Error (E_ERROR): gd-webp cannot allocate temporary buffer`) but maybe thats another issue .. I don't know.

**Comment by MrPetovan** [comment]
What PHP version are you running? If it's lower than PHP 8, GD doesn't have support for the WEBP format yet,

**Comment by haheute** [comment]
8.1.15

**Comment by MrPetovan** [comment]
Too bad, it was a good try. Not sure where the error comes from then.

**Comment by haheute** [comment]
Now it is again running with `memory_limit = 2048M` and I see these errors (instead of the "allocate error") all the time:
(with PHP 8.1.20)

![friendica-worker-killed](https://github.com/friendica/friendica/assets/5753419/6eacff28-12c0-490c-aab9-87d674f7bec4)

The command seems to be "SpoolPost" most of the time, plus some "Cron" and "Contact\Remove".

**Comment by bkil** [comment]
One seems to be known with an easy workaround:
https://stackoverflow.com/questions/69077798/trap-gd-webp-cannot-allocate-temporary-buffer-error-in-php-8-when-using-imagecre
https://github.com/Intervention/image/issues/1196

The original stack trace pointed to an SQL query returning too many rows to fit in memory 
https://github.com/friendica/friendica/blob/6fe85d9e2c0291d1fa68a89872c6a96962806d87/src/Database/Database.php#L547

**Comment by bkil** [comment]
The timeout error is controlled by the system => worker_max_duration config value. It can be down to 10 seconds in dozens of cases, including SpoolPost, Contact\Unfollow, Contact\RevokeFollow and more. This actually sounds pretty tight with a busy database or too many of these tasks in parallel.

* https://github.com/friendica/friendica/blob/ac1ebac112f36fa1685904cf61e24b7ad4008232/static/defaults.config.php#L592
* https://github.com/friendica/friendica/blob/ac1ebac112f36fa1685904cf61e24b7ad4008232/src/Core/Worker/Cron.php#L113

Although, it would be great to get to the bottom of this to ensure there are no slow unindexed SQL queries that stall you down.

**Comment by bkil** [comment]
It would be great to get a detailed/debug trace to see _who_ is calling this function or to find out the query from mysql.

https://stackoverflow.com/questions/9243070/why-allowed-memory-size-exhausted
https://www.php.net/manual/en/mysqlinfo.concepts.buffering.php

**Comment by haheute** [comment]
Can you tell me how to get such a debug trace? 

I would like to do something to help fix this, but I don't know what.. :-/

**Comment by haheute** [comment]
This happens, when I increase the memory limit from 256M to 512M:

![256-to-512-gimped](https://github.com/friendica/friendica/assets/5753419/234fe36c-4a61-46f5-99ed-694469b1de75)

The value, it tries to allocate, increases, too.

**Comment by annando** [comment]
Have you activated the decoupled receiver?

**Comment by haheute** [comment]
Yes, in the admin area..

**Comment by annando** [comment]
What happens, when you deactivate it?

**Comment by haheute** [comment]
It looks the same in the log view section. A new entry about every 5 minutes with:  `Fatal Error (E_ERROR): Allowed memory size of 536870912 bytes exhausted (tried to allocate 94388224 bytes)`

**Comment by annando** [comment]
Does this then occur in the worker as well?

**Comment by haheute** [comment]
Yes it says "worker":

`2024-01-15T08:24:15Z 	ALERT 	worker 	Fatal Error (E_ERROR): Allowed memory size of 536870912 bytes exhausted (tried to allocate 94388224 bytes)`

And most of the time this info:

file:   `/var/www/localhost/friendica/src/Database/Database.php`
line:   `568`

The trace field is empty.

**Comment by annando** [comment]
When it is caused by the database, then the result of a query contains too much data. Do you have a callstack?

**Comment by haheute** [comment]
I don't know what a callstack is. I can try to deliver one. Can you tell me how?

**Comment by haheute** [comment]
In the friendica log, such an error looks like this: 

`2024-01-16T08:57:09Z worker [ALERT]: Fatal Error (E_ERROR): Allowed memory size of 536870912 bytes exhausted (tried to allocate 94388224 bytes) {"code":1,"message":"Allowed memory size of 536870912 bytes exhausted (tried to allocate 94388224 bytes)","file":"/var/www/localhost/friendica/src/Database/Database.php","line":545,"trace":null} - {"file":null,"line":null,"function":null,"request-id":"65a644da6d610","stack":"ErrorHandler::handleFatalError","uid":"8ffa41","process_id":15621}`

**Comment by haheute** [comment]
I wonder why nobody else seem to have these errors.. I must be doing something wrong with my Alpine Linux. 
Since some days, there is also a `gd-webp`message (again). 

![gd-webp](https://github.com/friendica/friendica/assets/5753419/b28740a5-189e-437b-969c-a8ce7496b48f)

**Comment by AlfredSK** [comment]
I have those errors about allowed memory exhaustion too. Here they are related to Smileys.php. To be honest I simply ignore them. 🙂

**Comment by haheute** [comment]
I wonder why nobody else seems to have the `gd-webp` errors. I see them all the time.
It seems as if they appear together with the other error, like this:

```
2024-06-10T05:30:33Z 	ALERT 	worker 	Fatal Error (E_ERROR): Allowed memory size of 536870912 bytes exhausted (tried to allocate 96485376 bytes)
2024-06-10T05:30:29Z 	ALERT 	worker 	Fatal Error (E_ERROR): gd-webp cannot allocate temporary buffer
```
Currently I have set `memory_limit` to 512M.

Something must be wrong. What is the worker doing with images so often? Avatar images maybe? But how can a small avatar image exceed the memory limit?

**Comment by bkil** [comment]
I think once you have that animated webp stuck in your queue, it always gets back in due to the exception.

**Comment by haheute** [comment]
Thanks.. Is there a way to identify it and maybe remove it from the database?

**Comment by elrido** [comment]
Having now also been struck by this fatal error and not having figured out how to detect which message in the queue causes this, I resorted to ignoring webp images altogether, by adding this into Object/Image.php just before line 230:

```php
// [...]
               if ($this->originType == IMAGETYPE_WEBP) {        
                        return false;                            
                }                                               
                try {                                                
                        $this->image = @imagecreatefromstring($data);
// [...]
```

I would have preferred to only ignore animated ones (using also `&& $this->isAnimatedWebP(substr($data, 0, 90)))`), but in my case it seems the issue is not only with the animated ones or my case used some other unsupported properties. For debugging this further and actually solving the issue, it would be very valuable if we could extract one of the offending images and share it here for further investigations.

I've reverted my workaround and if I catch another one, I would gladly extract it from the database for this purpose. Which table(s) store(s) these unprocessed messages and might contain column(s) with unprocessed images to scan for webp's?

For reference, I'm using the docker image friendica:2024.03-fpm-alpine & PHP 8.1.27.

**Comment by haheute** [comment]
The problem seems to be solved for now, with the help of [this website](https://www.accuwebhosting.com/blog/php-workers-vs-memory-limit/).

I have set `pm = static` and `pm.max_children = 5` value in `/etc/php8.4/php-fpm.d/www.conf`. 
Then in the admin page in friendica,  the maximum number of workers, has been changed from `10` to `5`. 

The PHP `memory_limit` was set to `512M` (`536870912 bytes`) and the error message said, the memory was exhausted, and tried to allocate more memory. 

    2026-02-08T23:58:09Z worker [ALERT]: Fatal Error (E_ERROR): Allowed memory size of 536870912 bytes exhausted (tried to allocate 109068288 bytes) {"code":1,"message":"Allowed memory size of 536870912 bytes exhausted (tried to allocate 109068288 bytes)","file":"/opt/friendica/friendica/src/Database/Database.php","line":543,"trace":null,"worker_id":"17dfdaa","worker_cmd":null} - {"file":null,"line":null,"function":null,"request-id":"698922e1a0956","stack":"","uid":"aca26e","process_id":7441}
    2026-02-09T00:03:49Z worker [ALERT]: Fatal Error (E_ERROR): Allowed memory size of 536870912 bytes exhausted (tried to allocate 94388224 bytes) {"code":1,"message":"Allowed memory size of 536870912 bytes exhausted (tried to allocate 94388224 bytes)","file":"/opt/friendica/friendica/src/Database/Database.php","line":543,"trace":null,"worker_id":"dee316e","worker_cmd":null} - {"file":null,"line":null,"function":null,"request-id":"698924382f76f","stack":"","uid":"f3df53","process_id":7713}



Then I tried with `memory_limit = 1024M`, and the alert still appeared in the logs, every 5 minutes. The memory was still exhausted, but it tried to allocate a little less, now (`16384 bytes`).

    2026-02-09T00:10:44Z worker [ALERT]: Fatal Error (E_ERROR): Allowed memory size of 1073741824 bytes exhausted (tried to allocate 16384 bytes) {"code":1,"message":"Allowed memory size of 1073741824 bytes exhausted (tried to allocate 16384 bytes)","file":"/opt/friendica/friendica/src/Database/Database.php","line":543,"trace":null,"worker_id":"74b40d0","worker_cmd":null} - {"file":null,"line":null,"function":null,"request-id":"6989257ba784c","stack":"","uid":"3c8651","process_id":7916}

After setting `memory_limit = 2048M` the error/alert in the logs didn't appear again, since last night. Some worker task seemed to run on 100% CPU for some minutes and I then it looked as if the problem was finally solved.  

I still have set `memory_limit = 2048M`. If I understand correctly, with 5 workers, each one using maximum `2048M` the maximum would be about 10G. The server has only 10G of RAM. Maybe I should only use 4 workers. Or use some memory_limit between 1024 and 2048.
I heard, other friendica instances use only `memory_limit = 256M`. Maybe I should try that, too. But what, if again, such a big video (or what it was), comes..? I'll try.. maybe later.. maybe with `512M` again.. not sure..
